ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+semian_cb_data_set_successes(VALUE self, VALUE num)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  if (TYPE(num) != T_FIXNUM && TYPE(num) != T_FLOAT/*|| TYPE(size) != T_BIGNUM*/)
+    return Qnil;
+
+  ptr->shm_address->successes = NUM2INT(num);
+
+  semian_cb_data_unlock(self);
+  return num;
+}
+
+
+static VALUE
+semian_cb_data_semid(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  return INT2NUM(ptr->semid);
+}
+static VALUE
+semian_cb_data_shmid(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  return INT2NUM(ptr->shmid);
+}
+
+static VALUE
+semian_cb_data_array_at_index(VALUE self, VALUE idx)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (TYPE(idx) != T_FIXNUM && TYPE(idx) != T_FLOAT/*|| TYPE(size) != T_BIGNUM*/)
+    return Qnil;
+
+  int index = NUM2INT(idx);
+
+  if (index <0 || index >= ptr->arr_max_size) {
+    return Qnil;
+  }
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+  shared_cb_data *data = ptr->shm_address;
+  VALUE retval = index < (data->arr_length) ? DBL2NUM(data->errors[index]) : Qnil;
+
+  semian_cb_data_unlock(self);
+  return retval;
+
+}
+
+static VALUE
+semian_cb_data_array_set_index(VALUE self, VALUE idx, VALUE val)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  // check shared memory for NULL
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (TYPE(idx) != T_FIXNUM && TYPE(idx) != T_FLOAT/*|| TYPE(size) != T_BIGNUM*/)
+    return Qnil;
+  if (TYPE(val) != T_FIXNUM && TYPE(val) != T_FLOAT/*|| TYPE(size) != T_BIGNUM*/)
+    return Qnil;
+
+  int index = NUM2INT(idx);
+  double value = NUM2DBL(val);
+
+  if (index <0 || index >= ptr->arr_max_size) {
+    return Qnil;
+  }
+
+  if (!semian_cb_data_lock(self)){
+    return Qnil;
+  }
+
+  ptr->shm_address->errors[index] = value;
+  ptr->shm_address->arr_length = index+1;
+
+  semian_cb_data_unlock(self);
+  return val;
+
+}
+
+static VALUE
+semian_cb_data_array_length(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+  int arr_length =ptr->shm_address->arr_length;
+  semian_cb_data_unlock(self);
+  return INT2NUM(arr_length);
+}
+
+static VALUE
+semian_cb_data_set_push_back(VALUE self, VALUE num)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+  if (TYPE(num) != T_FIXNUM && TYPE(num) != T_FLOAT/*|| TYPE(size) != T_BIGNUM*/)
```

Comment:
Is the issue the precision/error of floats? In code the `FLOAT` is actually a C `double`. Otherwise, `(Time.now.to_f*1000).to_i` gives me `1443539804488` or ~ `1.4 E12`, which is a few orders of magnitude bigger than `2^31-1`, `2147483647` ~ `2.1 E9` the usual size of an `int`. 

So, issue is whether accuracy in ms is needed using `long`.